astropy/units/spectral_quantity.py

```diff
@@ -0,0 +1,291 @@
+from .quantity import SpecificTypeQuantity
+from .decorators import quantity_input
+from astropy.constants import c
+
+__all__ = ['SpectralQuantity']
+
+# We don't want to run doctests in the docstrings we inherit from Quantity
+__doctest_skip__ = ['SpectralQuantity.*']
+
+
+SPECTRAL_UNITS = (si.Hz, si.m, si.J, (1 / si.m).unit, si.m / si.s)
+
+DOPPLER_CONVENTIONS = {
+    'radio': eq.doppler_radio,
+    'optical': eq.doppler_optical,
+    'relativistic': eq.doppler_relativistic
+}
+
+
+class SpectralQuantity(SpecificTypeQuantity):
+    """
+    One or more value(s) with spectral units.
+
+    The spectral units should be those for frequencies, wavelengths, energies,
+    wavenumbers, or velocities (interpreted as Doppler velocities relative to a
+    rest spectral value). The advantage of using this class over the regular
+    `~astropy.units.Quantity` class is that in `SpectralQuantity`, the
+    ``u.spectral`` equivalency is enabled by default (allowing automatic
+    conversion between spectral units), and a preferred Doppler rest value and
+    convention can be stored for easy conversion to/from velocities.
+
+    Parameters
+    ----------
+    value : ndarray or `~astropy.units.Quantity` or `SpectralQuantity`
+        Spectral axis data values.
+    unit : str or `~astropy.units.Unit`
+        Unit for the given data.
+    doppler_rest : `~astropy.units.Quantity`, optional
+        The rest value to use for conversions from/to velocities
+    doppler_convention : str, optional
+        The convention to use when converting the spectral data to/from
+        velocities.
+    """
+
+    _equivalent_unit = SPECTRAL_UNITS
+
+    _include_easy_conversion_members = True
+
+    def __new__(cls, value, unit=None,
+                doppler_rest=None, doppler_convention=None,
+                **kwargs):
+
+        kwargs['subok'] = True
+
+        obj = super().__new__(cls, value, unit=unit, **kwargs)
+
+        # The quantity machinery will drop the unit because type(value) !=
+        #  SpectralQuantity when passing in a Quantity object. Reassign the unit
+        #  here to avoid this.
+        # if isinstance(value, u.Quantity) and unit is None:
+        #     obj._unit = value.unit
+
+        # If we're initializing from an existing SpectralQuantity, keep any
+        # parameters that aren't being overridden
+        if isinstance(value, SpectralQuantity):
+            if doppler_rest is None:
+                doppler_rest = value.doppler_rest
+            if doppler_convention is None:
+                doppler_convention = value.doppler_convention
+
+        obj._doppler_rest = doppler_rest
+        obj._doppler_convention = doppler_convention
+
+        return obj
+
+    def __array_finalize__(self, obj):
+        super().__array_finalize__(obj)
+        self._doppler_rest = getattr(obj, '_doppler_rest', None)
+        self._doppler_convention = getattr(obj, '_doppler_convention', None)
+
+    def __quantity_subclass__(self, unit):
```

Comment:
Ok so this is an interesting problem - AA/s is actually a valid spectral unit since it's velocity! For now I've removed ``__quantity_subclass__`` but see the ``test_arithmetic`` test for SpectralQuantity - there is a commented out line that fails because of this. This is beyond by units-fu - what would be the correct way to make it so that changes in unit only work if done through a ``to`` conversion but not arithmetic?

Reply:
Ouch...  For now maybe easiest to just rely on what `SpecificTypeQuantity` does and we can try to fix this later.